lib/hmmlearn/base.py

```diff
@@ -494,20 +561,23 @@ def fit(self, X, lengths=None):
             "scaling": self._fit_scaling,
             "log": self._fit_log,
         }[self.implementation]
-        for iter in range(self.n_iter):
-            stats = self._initialize_sufficient_statistics()
-            curr_logprob = 0
-            for sub_X in _utils.split_X_lengths(X, lengths):
-                lattice, logprob, posteriors, fwdlattice, bwdlattice = \
-                        impl(sub_X)
-                # Derived HMM classes will implement the following method to
-                # update their probability distributions, so keep
-                # a single call to this method for simplicity.
+        
+        def process_sequence(sub_X, stats):
+            lattice, logprob, posteriors, fwdlattice, bwdlattice = \
+                    impl(sub_X)
+            # Derived HMM classes will implement the following method to
+            # update their probability distributions, so keep
+            # a single call to this method for simplicity.
+            with self.stats_lock:
```

Comment:
The need for locking here vaguely suggests to me that _accumulate_sufficient_statistics is not, in fact, the right API; rather, we should perhaps have a _summarize_sufficient_statistics that *returns* the info for a chunk and then aggregate the chunks' individual infos in _do_mstep.  Thoughts?
(Perhaps also relevant for @dizcza wrt #437?)

Reply:
Yes it would be great to get rid of the lock here. Returning the info and aggregating it later would be one way, especially if the aggregation is order independent. However looking at some of the different HMM implementations, there's quite a lot data involved which might lead to increased memory usage.

Reply:
IMHO the aggregation should be order-independent. But as @Dobatymo pointed out, the unreduced statistics gathered from a batch of samples could be quite heavy in memory usage, and transferring it here and there is not a good idea (again, IMHO). The approach I want to bring on top of #437 is the ability to reduce (to do a summation over the batch dimension) directly in the `_accumulate_sufficient_statistics` function. Therefore, a lock is needed. But after the summation is done - no need to wait while each thread/process does its summation - they are independent. And I agree with @anntzer that we'll benefit from having a `_summarize_sufficient_statistics` function, in which each job reduces the batch it processes (does the summation) and returns the summary that can be efficiently added to the global accumulated statistics.
Just like in CUDA programming.